src/main/java/com/github/shyiko/mysql/binlog/event/deserialization/BinaryLogEventDataReader.java

```diff
@@ -0,0 +1,317 @@
+        return bytes;
+    }
+
+    public void skip(int length) {
+        buffer.position(buffer.position() + length);
+    }
+
+    /**
+     * Reads packed long value from the current position in the buffer.
+     * <p>
+     * Format (first-byte-based):<br>
+     * 0-250 - The first byte is the number (in the range 0-250). No additional bytes are used.<br>
+     * 251 - SQL NULL value<br>
+     * 252 - Two more bytes are used. The number is in the range 251-0xffff.<br>
+     * 253 - Three more bytes are used. The number is in the range 0xffff-0xffffff.<br>
+     * 254 - Eight more bytes are used. The number is in the range 0xffffff-0xffffffffffffffff.
+     * @return long or null
+     * @throws IOException if unexpected bytes are encountered (IOException is used for compatibility with
+     *                     {@link com.github.shyiko.mysql.binlog.io.ByteArrayInputStream}.
+     */
+    public long readPackedLong() throws IOException {
+        int b = (buffer.get() & 0xFF);
+        if (b < 251) {
+            return b;
+        } else if (b == 251) {
+            throw new IOException("Unexpected NULL value");
+        } else if (b == 252) {
+            return (buffer.getShort() & 0xFFFFL);
+        } else if (b == 253) {
+            return (buffer.get() & 0xFFL) | ((buffer.get() & 0xFFL) << 8) | ((buffer.get() & 0xFFL) << 16);
+        } else if (b == 254) {
+            return buffer.getLong();
+        }
+        throw new RuntimeException("Unexpected packed number byte " + b);
+    }
+
+    /**
+     * Reads packed integer value from the current position in the buffer. Similar to {@link #readPackedLong()} but
+     * makes sure that the number read is fit into integer.
+     * @throws IOException if unexpected bytes are encountered or in case of integer overflow (IOException is used for
+     *                      compatibility with {@link com.github.shyiko.mysql.binlog.io.ByteArrayInputStream}.
+     */
+    public int readPackedInteger() throws IOException {
+        long value = readPackedLong();
+        if (value > Integer.MAX_VALUE || value < Integer.MIN_VALUE) {
+            throw new IOException("Packed integer " + value + " is out of Integer range");
+        }
+        return (int)value;
+    }
+
+
+    public BitSet readBitSet(int length, boolean bigEndian) {
+        // According to MySQL internals the amount of storage required for N columns is INT((N+7)/8) bytes.
+        int lengthInBytes = (length + 7) >> 3;
+        byte[] bytes = new byte[lengthInBytes];
+        buffer.get(bytes);
+
+        // BitSet.valueOf we're using below interprets bytes as little-endian
+        if (!bigEndian) {
+            bytes = reverse(bytes);
+        }
+
+        // Zero out not used bits.
+        byte mask = (byte)(0xff >> (bytes.length * 8 - length));
+        bytes[bytes.length - 1] = (byte)(bytes[bytes.length - 1] & mask);
+
+        return BitSet.valueOf(bytes);
+    }
+
+    public int available() {
```

Comment:
I'm having trouble understanding the `blockLength` in the ByteArrayInputStream version of `available()`. Do you have some more insight there? It seems in the new data reader version, the buffer is the "block" and we resize the buffer limit in `enterBlock`. Whereas in the old input stream version, it seems a bit more hacky and has to block off portions of the input stream?